src/lib/environment/lexical_environment.rs

```diff
@@ -98,25 +107,78 @@ impl LexicalEnvironment {
             .get_global_object()
     }
 
-    pub fn create_mutable_binding(&mut self, name: String, deletion: bool) {
-        self.get_current_environment()
-            .borrow_mut()
-            .create_mutable_binding(name, deletion)
+    pub fn create_mutable_binding(&mut self, name: String, deletion: bool, scope: VariableScope) {
+        match scope {
+            VariableScope::Block => self
+                .get_current_environment()
+                .borrow_mut()
+                .create_mutable_binding(name, deletion),
+            VariableScope::Function => {
+                // Find the first function or global environment (from the top of the stack)
+                let env = self
+                    .environment_stack
```

Comment:
This is great, although worth pointing out.
I want to move away from using the stack at some point.

Instead every environment has an `outer_env` pointer to the environment above them
https://github.com/jasonwilliams/boa/blob/master/src/lib/environment/declarative_environment_record.rs#L167-L169

It will be more flexible in future if we can rely on this.
The stack will only get us so far, but traversing an environment tree should rely on traversing it's parent_env until you reach null.

Reply:
Trait methods for every environment record:
https://github.com/jasonwilliams/boa/blob/master/src/lib/environment/environment_record_trait.rs#L71-L75

Reply:
Ah okay, then I will refactor the code to reflect this. Perhaps I could add a function to the trait which returns an iterator over the parent environments?

Reply:
@barskern i don't think you need that, there's only 1 place that's going to be iterating up the environments and that's here:
https://github.com/jasonwilliams/boa/blob/master/src/lib/environment/lexical_environment.rs#L141-L173

Setting here:
https://github.com/jasonwilliams/boa/blob/master/src/lib/environment/lexical_environment.rs#L113-L115

environments themselves shouldn't care about iterating (that's the job of the `lexical_environment` module